src/library/autodj/autodjprocessor.cpp

```diff
@@ -211,30 +217,75 @@ void AutoDJProcessor::fadeNow() {
-        leftDeck.setRepeat(false);
-    } else if (rightDeck.isPlaying()) {
-        // Make sure rightDeck.fadeDuration is up to date.
-        calculateTransition(&rightDeck, &leftDeck);
-
-        // override posThreshold to start fade now
-        rightDeck.posThreshold = rightDeck.playPosition() -
-                ((1.0 - crossfader) / 2 * (rightDeck.fadeDuration));
-        // Repeat is disabled by FadeNow but disables auto Fade
-        rightDeck.setRepeat(false);
-    }
-    //else {
-    //    No deck playing, do not know what to do
+    DeckAttributes* pLeftDeck = m_decks[0];
+    DeckAttributes* pRightDeck = m_decks[1];
+    DeckAttributes* pFromDeck;
+    DeckAttributes* pToDeck;
+
+    if (pLeftDeck->isPlaying() &&
+            (!pRightDeck->isPlaying() || crossfader < 0.0)) {
+        pFromDeck = pLeftDeck;
+        pToDeck = pRightDeck;
+    } else if (pRightDeck->isPlaying()) {
+        pFromDeck = pRightDeck;
+        pToDeck = pLeftDeck;
+    } else {
+        // Both decks are playing or no decks are playing;
+        // fading now makes no sense.
+        return;
+    }
+
+    pFromDeck->setRepeat(false);
+    pFromDeck->isFromDeck = true;
+    pToDeck->isFromDeck = false;
+
+    double fromDeckCurrentPosition = pFromDeck->playPosition() * pFromDeck->duration();
+    double toDeckCurrentPosition = pToDeck->playPosition() * pToDeck->duration();
+
+    pFromDeck->fadeBeginPos = fromDeckCurrentPosition;
+    // Do not seek to a calculated start point; start the to deck from wherever
+    // it is if the user has seeked since loading the track.
+    pToDeck->startPos = toDeckCurrentPosition;
+
+    // If the user presses "Fade now", assume they want to fade *now*, not later.
+    // So if the spinbox time is negative, do not insert silence.
+    double spinboxTime = fabs(m_transitionTime);
+
+    double fadeTime;
+    if (m_transitionMode == TransitionMode::FullIntroOutro
+       || m_transitionMode == TransitionMode::AlignIntroOutroStart) {
+        // Use the intro length as the transition time. If the user has seeked
+        // away from the intro start since the track was loaded, start from
+        // there and do not seek back to the intro start. If they have seeked
+        // past the introEnd or the introEnd is not marked, fall back to the
+        // spinbox time.
+        double outroEnd = getOutroEndPosition(pFromDeck);
+        double introEnd = getIntroEndPosition(pToDeck);
+
+        if (toDeckCurrentPosition <= introEnd && introEnd >= 0) {
+            double introLength = introEnd - toDeckCurrentPosition;
+            if (outroEnd >= 0) {
+                // The fade must end by the outro end at the latest.
+                double timeUntilOutroEnd = outroEnd - fromDeckCurrentPosition;
+                fadeTime = math_min(introLength, timeUntilOutroEnd);
+            } else {
+                fadeTime = introLength;
+            }
+        } else {
+            fadeTime = spinboxTime;
```

Comment:
math_min(spinboxTime, timeUntilOutroEnd);
Even in this case we need to respect the timeUntilOutroEnd, because the spinboxTime will not have already triggered the fade if negative.